### PR TITLE
Chapter 1 - Introduction

### PR DESCRIPTION
**Pull request by Martin521**
This is chapter 1 (Introduction)
Please review and merge the infrastructure first, then the chapters.

### CODE REVIEW

**Review by vzarytovskii — APPROVED** [review]
Finally got to it. Looks good, thanks!

**Comment by edgarfgp** [comment]
Maybe we can ask @abelbraaksma to help us review all this PR’s.

**Comment by vzarytovskii** [comment]
Already did on slack, he will get to it once he has time for it

File: spec/introduction.md [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+    let htmlOfSites =
+        Async.Parallel [for site in sites -> http site ]
+        |> Async.RunSynchronously
+```
+
+By using asynchronous workflows together with other CLI libraries, F# programs can implement parallel tasks, parallel I/O operations, and message-receiving agents.
+
+### Strong Typing for Floating-Point Code
+
+F# applies type checking and type inference to floating-point-intensive domains through units of measure inference and checking. This feature allows you to type-check programs that manipulate floating-point numbers that represent physical and abstract quantities in a stronger way than other typed languages, without losing any performance in your compiled code. You can think of this feature as providing a type system for floating-point code.
```
Comment (abelbraaksma) [inline spec/introduction.md:257]:
```suggestion
F# applies type checking and type inference to floating-point-intensive domains through *units of measure inference and checking*. This feature allows you to type-check programs that manipulate floating-point numbers that represent physical and abstract quantities in a stronger way than other typed languages, without losing any performance in your compiled code. You can think of this feature as providing a type system for floating-point code.
```

File: spec/introduction.md [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    let gravityOnEarth = 9.81<m/s^2>
+    let heightOfTowerOfPisa = 55.86<m>
+    let speedOfImpact = sqrt(2.0 * gravityOnEarth * heightOfTowerOfPisa)
+```
+
+The `Measure` attribute tells F# that `kg`, `s`, and `m` are not really types in the usual sense of the word, but are used to build units of measure. Here `speedOfImpact` is inferred to have type `float<m/s>`.
+
+### Object-Oriented Programming and Code Organization
+
+The sample program shown at the start of this chapter is a _script_. Although scripts are excellent for rapid prototyping, they are not suitable for larger software components. F# supports the transition from scripting to structured code through several techniques.
```
Comment (abelbraaksma) [inline spec/introduction.md:275]: We can use single style for markdown in such a large body of work (VSCode can do this), and we can make this a lint-check in PRs as well perhaps? Not sure how precise we should be here.
```suggestion
The sample program shown at the start of this chapter is a *script*. Although scripts are excellent for rapid prototyping, they are not suitable for larger software components. F# supports the transition from scripting to structured code through several techniques.
```

File: spec/introduction.md [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+Other mechanisms aimed at supporting software engineering include *signatures*, which can be used to give explicit types to components, and namespaces, which serve as a way of organizing the name hierarchies for larger APIs.
+
+### Information-rich Programming
+
+F# Information-rich programming addresses the trend toward greater availability of data, services, and information. The key to information-rich programming is to eliminate barriers to working with diverse information sources that are available on the Internet and in modern enterprise environments. Type providers and query expressions are a significant part of F# support for information-rich programming.
+
+The F# Type Provider mechanism allows you to seamlessly incorporate, in a strongly typed manner, data and services from external sources. A type provider presents your program with new types and methods that are typically based on the schemas of external information sources. For example, an F# type provider for Structured Query Language (SQL) supplies types and methods that allow programmers to work directly with the tables of any SQL database:
```
Comment (abelbraaksma) [inline spec/introduction.md:365]:
```suggestion
The F# Type Provider mechanism allows you to seamlessly incorporate, in a strongly typed manner, data and services from external sources. A *type provider* presents your program with new types and methods that are typically based on the schemas of external information sources. For example, an F# type provider for Structured Query Language (SQL) supplies types and methods that allow programmers to work directly with the tables of any SQL database:
```

File: spec/introduction.md [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+
+Regular expressions are given in the usual notation, as shown in the table:
+
+| Notation | Meaning |
+| --- | --- |
+| regexp+ | One or more occurrences |
+| regexp* | Zero or more occurrences |
+| regexp? | Zero or one occurrences |
+| [ char - char ] | Range of ASCII characters |
+| [ ^ char - char ] | Any characters except those in the range |
```
Comment (abelbraaksma) [inline spec/introduction.md:401]: Code readability (in VSCode, I use 'Markdown all in one' to format a block)
```suggestion
| Notation          | Meaning                                  |
| ----------------- | ---------------------------------------- |
| regexp+           | One or more occurrences                  |
| regexp*           | Zero or more occurrences                 |
| regexp?           | Zero or one occurrences                  |
| [ char - char ]   | Range of ASCII characters                |
| [ ^ char - char ] | Any characters except those in the range |
```

File: spec/introduction.md [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+
+Unicode character classes are referred to by their abbreviation as used in CLI libraries for regular expressions—for example, `\Lu` refers to any uppercase letter. The following characters are referred to using the indicated notation:
+
+| Character | Name | Notation |
+| --- | --- | --- |
+| \b | backspace | ASCII/UTF-8/UTF-16/UTF-32 code 08 |
+| \n | newline | ASCII/UTF-8/UTF-16/UTF-32 code 10 |
+| \r | return | ASCII/UTF-8/UTF-16/UTF-32 code 13 |
+| \t | tab | ASCII/UTF-8/UTF-16/UTF-32 code 09 |
```
Comment (abelbraaksma) [inline spec/introduction.md:410]: Same:
```suggestion
| Character | Name      | Notation                          |
| --------- | --------- | --------------------------------- |
| \b        | backspace | ASCII/UTF-8/UTF-16/UTF-32 code 08 |
| \n        | newline   | ASCII/UTF-8/UTF-16/UTF-32 code 10 |
| \r        | return    | ASCII/UTF-8/UTF-16/UTF-32 code 13 |
| \t        | tab       | ASCII/UTF-8/UTF-16/UTF-32 code 09 |
```

File: spec/introduction.md [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+
+This regular expression matches precisely the characters `#if`:
+
+    "#if"
+
+Regular expressions are typically used to specify tokens.
+
+    token token-name = regexp
+
+In the grammar rules, the notation `element-name~opt` indicates an optional element. The notation `...` indicates repetition of the preceding non-terminal construct and the separator token. For example, `expr ',' ... ',' expr` means a sequence of one or more `expr` elements separated by commas.
```
Comment (abelbraaksma) [inline spec/introduction.md:430]: maybe:
```suggestion
In the grammar rules, the notation `element-nameₒₚₜ` indicates an optional element. The notation `...` indicates repetition of the preceding non-terminal construct and the separator token. For example, `expr ',' ... ',' expr` means a sequence of one or more `expr` elements separated by commas.
```

**Review by abelbraaksma — CHANGES_REQUESTED** [review]
Mostly just very minor suggestions related to (small) differences with the original text. Plus some suggestions w.r.t. layout/markup/style here and there:

**Comment by abelbraaksma** [comment]
If we are in agreement with these changes, I can also just make them for you, but I think that requires a separate PR, as I don't have access to your forked branch.

File: spec/introduction.md [hunk 7]
```diff
@@ -0,0 +1,430 @@
+# Introduction
```
[the remark this answers is not in the document]
Comment (Martin521) [reply spec/introduction.md:1]: It is special static text, see [spec/front-matter.md](https://github.com/fsharp/fslang-spec/blob/main/spec/front-matter.md), which is merged already. (But may still need polishing.)

File: spec/introduction.md [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+F# is a scalable, succinct, type-safe, type-inferred, efficiently executing functional/imperative/object-oriented programming language. It aims to be the premier typed functional programming language for the .NET framework and other implementations of the Ecma 335 Common Language Infrastructure (CLI) specification. F# was partly inspired by the OCaml language and shares some common core constructs with it.
```
[the remark this answers is not in the document]
Comment (vzarytovskii) [reply spec/introduction.md]: I think we want to first merge as is, and then make a branch/tag with new more fresh version, so there are no surprises if we want to review changes, etc.

File: spec/introduction.md (hunk 2 above)
Comment (Martin521) [reply spec/introduction.md:274]: I agree, we should use a single style

File: spec/introduction.md [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+results:
+
+    > let vowels = ['e'; 'i'; 'o'; 'u'];;
+    val vowels: char list = ['e'; 'i'; 'o'; 'u']
+
+    > ['a'] @ vowels;;
+    val it: char list = ['a'; 'e'; 'i'; 'o'; 'u']
+    
+    > vowels @ ['y'];;
+    val it: char list = ['e'; 'i'; 'o'; 'u'; 'y']
```
[the remark this answers is not in the document]
Comment (Martin521) [reply spec/introduction.md]: I agree it looks nicer as code block, but we have reserved the `fsharp` info string for F# code that we could later extract and verify (like the C# spec does or at least did it).
See also the [conversion guideline](urlhttps://github.com/amplifying-fsharp/fsharplang) that we used.
I will later go separately through the code and turn the fsi examples into code blocks with the `fsother` info string. (As, I believe, I did in later chapters.)

File: spec/introduction.md [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+        "http://www.bing.com"
+        "http://www.google.com"
+        "http://www.yahoo.com"
+        "http://www.search.com"
+        ]
+
+    let htmlOfSites =
+        Async.Parallel [for site in sites -> http site ]
+        |> Async.RunSynchronously
+```
```
[the remark this answers is not in the document]
Comment (Martin521) [reply spec/introduction.md:252]: Yes, Fantomas makes sense. But I would like to do that then in an automated way. See also my above remark of using code blocks with the `fsharp` info string only for F# code.

**Comment by Martin521** [comment]
> If we are in agreement with these changes, I can also just make them for you, but I think that requires a separate PR, as I don't have access to your forked branch.

Thanks for the thorough review. I will look at the "opt" issue next.

**Comment by abelbraaksma** [comment]
@Martin521, great that you're tackling this. About b6d8fd4 and the like: I didn't create a comment for each occurrence of "top-level code should not be indented". This is probably something that ought to be changed for **all code occurrences**, across the PRs. 

Did you use a script for this? Cause then a change to that script would suffice, hopefully :).

**Comment by abelbraaksma** [comment]
wow, great work! I'll go over this one more time, see if we've caught everything (and otherwise, there's always another day, right?).

Btw, can we strategize about changes that aren't the same as the source? My take in this review is the following:

* Keep the 1:1 correspondence as much as possible
* Where there's a clear error or typo in the source (i.e., misspelling), then propose to change
* If the error is non-trivial or changes the wording significantly, propose a change separately (i.e. in a PR after merge) to gain input and consensus

I think, unless we stick to this, it would become a much too large effort in the first go (i.e. there were many places where I noticed mistakes in the original text, but I decided to let it slide as to not muddy the main conversion process).

File: spec/introduction.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (abelbraaksma) [reply spec/introduction.md:1]: for some reason, I cannot mark this resolved...

**Comment by abelbraaksma** [comment]
> I agree it looks nicer as code block, but we have reserved the fsharp info string for F# code that we could later extract and verify (like the C# spec does or at least did it).
See also the conversion guideline that we used.
I will later go separately through the code and turn the fsi examples into code blocks with the fsother info string. (As, I believe, I did in later chapters.)

On this point: the FSI examples are prime examples that can be auto-verified (i.e., parse the string after `>` and see if the result is the same as the shown output). But if we call it `fsother` or `fsi`, perhaps we can make that an alias for `f#` coloring (or, better, perhaps we can make `fsi` a proper tag so that it recognizes `>`? But that's a future scen)

File: spec/introduction.md (hunk 6 above)
Comment (abelbraaksma) [reply spec/introduction.md]: Per conclusion in that discussion: it is now a question mark (`?`).

File: spec/introduction.md [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+
+### Information-rich Programming
+
+F# Information-rich programming addresses the trend toward greater availability of data, services, and information. The key to information-rich programming is to eliminate barriers to working with diverse information sources that are available on the Internet and in modern enterprise environments. Type providers and query expressions are a significant part of F# support for information-rich programming.
+
+The F# Type Provider mechanism allows you to seamlessly incorporate, in a strongly typed manner, data and services from external sources. A *type provider* presents your program with new types and methods that are typically based on the schemas of external information sources. For example, an F# type provider for Structured Query Language (SQL) supplies types and methods that allow programmers to work directly with the tables of any SQL database:
+
+```fsharp
+// Add References to FSharp.Data.TypeProviders, System.Data, and System.Data.   Linq
+type schema = SqlDataConnection<"Data Source=localhost;Integrated   Security=SSPI;">
```
Comment (abelbraaksma) [inline spec/introduction.md:383]: nit:
```suggestion
// Add References to FSharp.Data.TypeProviders, System.Data, and System.Data.Linq
type schema = SqlDataConnection<"Data Source=localhost;Integrated Security=SSPI;">
```

File: spec/introduction.md [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+| Character | Name      | Notation                          |
+| --------- | --------- | --------------------------------- |
+| \b        | backspace | ASCII/UTF-8/UTF-16/UTF-32 code 08 |
+| \n        | newline   | ASCII/UTF-8/UTF-16/UTF-32 code 10 |
+| \r        | return    | ASCII/UTF-8/UTF-16/UTF-32 code 13 |
+| \t        | tab       | ASCII/UTF-8/UTF-16/UTF-32 code 09 |
+
+Strings of characters that are clearly not a regular expression are written verbatim. Therefore, the following string
+
+    abstract
```
Comment (abelbraaksma) [inline spec/introduction.md:428]: Not necessary now, but to recognize these blocks automatically, perhaps we can start using `abnf` sections (even though it isn't exactly abnf)?
`````suggestion
```abnf
abstract
```
`````

File: spec/introduction.md [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+
+matches precisely the characters `abstract`.
+
+Where appropriate, apostrophes and quotation marks enclose symbols that are used in the specification of the grammar itself, such as `'<'` and `'|'`. For example, the following regular expression matches `(+)` or `(-)`:
+
+    '(' (+|-) ')'
```
Comment (abelbraaksma) [inline spec/introduction.md:434]: Same here:
`````suggestion
```abnf
'(' (+|-) ')'
```
`````

File: spec/introduction.md [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+
+This regular expression matches precisely the characters `#if`:
+
+    "#if"
```
Comment (abelbraaksma) [inline spec/introduction.md:438]:
`````suggestion
```abnf
"#if"
```
`````

File: spec/introduction.md [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+
+Regular expressions are typically used to specify tokens.
+
+    token token-name = regexp
```
Comment (abelbraaksma) [inline spec/introduction.md:442]:
`````suggestion
```abnf
token token-name = regexp
```
`````

**Review by abelbraaksma — APPROVED** [review]
Looking good!!!

Now that I understand the post-conversion ideas a bit better, perhaps consider this for the (fake) abnf sections?
Oh, and one nit. Otherwise, good to go. I'll mark approved already, these aren't blockers.

**Comment by Martin521** [comment]
> You wrote ([#4 (comment)](https://github.com/fsharp/fslang-spec/pull/4#discussion_r1616901106)):
> 
> > I agree it looks nicer as code block, but we have reserved the fsharp info string for F# code that we could later extract and verify (like the C# spec does or at least did it).
> > See also the conversion guideline that we used.
> > I will later go separately through the code and turn the fsi examples into code blocks with the fsother info string. (As, I believe, I did in later chapters.)
> 
> On this point: the FSI examples are prime examples that can be auto-verified (i.e., parse the string after `>` and see if the result is the same as the shown output). But if we call it `fsother` or `fsi` (which can then still be auto-verified), perhaps we can make that an alias for `f#` coloring (or, better, perhaps we can make `fsi` a proper tag so that it recognizes `>`? But that's a future scen)

True. However, there are only the few fsi examples in this chapter and one more in one of the other chapters. So it won't be a big effort if we decide at some point to do something like this.

**Comment by Martin521** [comment]
> Looking good!!!
> 
> Now that I understand the post-conversion ideas a bit better, perhaps consider this for the (fake) abnf sections? It may help with any later automation we may need to do.
> 
> Oh, and one tiny nit. Otherwise, good to go. I'll mark approved already, these aren't blockers.

All the grammar pieces in the other chapters are marked 'fsgrammar' (which we can easiliy replace with something like abnf later if we want so). I did not mark these definitions here that way because it is not really F# grammar but sort of the grammar of the grammar (or, as the heading says, "notational conventions").

So, if it is ok for you, I will leave this as is.

**Comment by abelbraaksma** [comment]
Great work!  @vzarytovskii this is ready. I don't (yet) have the ability to merge, can you, or someone else with merge access, merge it?

**Pull request merged**